src/main/drivers/flash.c

```diff
@@ -47,13 +51,142 @@
+    int phase = TRY_1LINE;
+
+#ifdef USE_FLASH_MEMORY_MAPPED
+    bool memoryMappedModeEnabledOnBoot = isMemoryMappedModeEnabledOnBoot();
+#else
+    bool memoryMappedModeEnabledOnBoot = false
+#endif
+
+#ifndef USE_OCTOSPI_EXPERIMENTAL
+    if (!memoryMappedModeEnabledOnBoot) {
+        return false; // Not supported yet, enable USE_OCTOSPI_EXPERIMENTAL and test/update implementation as required.
+    }
+#endif
+
+    OCTOSPI_TypeDef *instance = octoSpiInstanceByDevice(OCTOSPI_CFG_TO_DEV(flashConfig->octoSpiDevice));
+
+    flashDevice.io.handle.octoSpi = instance;
+    flashDevice.io.mode = FLASHIO_OCTOSPI;
+
+    if (memoryMappedModeEnabledOnBoot) {
+        flashMemoryMappedModeDisable();
+    }
+
+    do {
+#ifdef USE_OCTOSPI_EXPERIMENTAL
+        if (!memoryMappedMode) {
+            octoSpiSetDivisor(instance, OCTOSPI_CLOCK_INITIALISATION);
+        }
+#endif
+        // for the memory-mapped use-case, we rely on the bootloader to have already selected the correct speed for the flash chip.
+
+        // 3 bytes for what we need, but some IC's need 8 dummy cycles after the instruction, so read 4 and make two attempts to
+        // assemble the chip id from the response.
+        uint8_t readIdResponse[4];
+
+        bool status = false;
+        switch (phase) {
+        case TRY_1LINE:
+            status = octoSpiReceive1LINE(instance, FLASH_INSTRUCTION_RDID, 0, readIdResponse, 4);
+            break;
+        case TRY_4LINE:
+            status = octoSpiReceive4LINES(instance, FLASH_INSTRUCTION_RDID, 2, readIdResponse, 3);
+            break;
+        default:
+            break;
+        }
+
+        if (!status) {
+            phase++;
+            continue;
+        }
+
+#ifdef USE_OCTOSPI_EXPERIMENTAL
+        if (!memoryMappedModeEnabledOnBoot) {
+            octoSpiSetDivisor(instance, OCTOSPI_CLOCK_ULTRAFAST);
+        }
+#endif
+
+        for (uint8_t offset = 0; offset <= 1 && !detected; offset++) {
+
+            uint32_t jedecID = (readIdResponse[offset + 0] << 16) | (readIdResponse[offset + 1] << 8) | (readIdResponse[offset + 2]);
+
+            if (offset == 0) {
+#if defined(USE_FLASH_W25Q128FV)
+                if (!detected && w25q128fv_identify(&flashDevice, jedecID)) {
+                    detected = true;
+                }
+#endif
+            }
+
+            if (offset == 1) {
+#ifdef USE_OCTOSPI_EXPERIMENTAL
+                if (!memoryMappedModeEnabledOnBoot) {
+                    // These flash chips DO NOT support memory mapped mode; suitable flash read commands must be available.
+#if defined(USE_FLASH_W25N01G)
+                    if (!detected && w25n01g_identify(&flashDevice, jedecID)) {
+                        detected = true;
+                    }
+#endif
+#if defined(USE_FLASH_W25M02G)
+                    if (!detected && w25m_identify(&flashDevice, jedecID)) {
+                        detected = true;
+                    }
+#endif
+                }
+#endif
+            }
```

Comment:
```suggestion
            uint32_t jedecID = (readIdResponse[0] << 24) | (readIdResponse[1] << 16) | (readIdResponse[2] << 8) | (readIdResponse[3]);
#if defined(USE_FLASH_W25Q128FV)
                if (w25q128fv_identify(&flashDevice, jedecID >> 8)) {
                    detected = true;
                    break;
                }
#endif
#ifdef USE_OCTOSPI_EXPERIMENTAL
                if (!memoryMappedModeEnabledOnBoot) {
                    // These flash chips DO NOT support memory mapped mode; suitable flash read commands must be available.
#if defined(USE_FLASH_W25N01G)
                    if (w25n01g_identify(&flashDevice, jedecID & 0xffffff)) {
                        detected = true;
                        break;
                    }
#endif
#if defined(USE_FLASH_W25M02G)
                    if (w25m_identify(&flashDevice, jedecID & 0xffffff)) {
                        detected = true;
                        break;
                    }
#endif
                }
#endif
            }
```

and possibly use macro/function for jedecID alignment

Reply:
this follows the pattern of the QSPI init code.  A later commit can address both at the same time, but for now they are consistent.